### ISSUE TITLE
Restic prune not removing data from B2 backend

### ISSUE THREAD

**Issue by hugoamvieira** [restic/restic#4421]
Output of `restic version`
--------------------------
`restic 0.15.2 compiled with go1.20.3 on linux/amd64`

How did you run restic exactly?
-------------------------------
I have a Proxmox VM backup script running in a cron job. When the script runs, it executes `vzdump` to perform the backup. `vzdump` then outputs the raw data to `stdout` and `restic` reads from `stdin`.

Here's what the backup command looks like:

```bash
vzdump <vm_id> --mode snapshot --compress zstd --stdout | restic backup --stdin --stdin-filename backup.vma.zst
```

After the backup is complete, the script executes a forget-prune operation with the specified snapshot keep policy, like this:

```bash
restic forget --prune \
  -d $KEEP_DAILY \
  -w $KEEP_WEEKLY \
  -m $KEEP_MONTHLY \
  -y $KEEP_YEARLY
```

This backup is a ~50GB file. I noticed that my Backblaze B2 bucket reported ~100GB of space. I checked the snapshots and there were two present. This made sense to me since the file itself is essentially a compressed and encrypted filesystem, so I expected each backup to essentially add 50GB worth of data to the remote.

I then changed the keep policy to `1` in all `KEEP_*` env vars and manually ran the forget-prune operation (same command as above), to only keep 1 snapshot. Here's the output of that:

```bash
repository N opened (version 2, compression level auto)
Applying Policy: keep 1 daily, 1 weekly, 1 monthly, 1 yearly snapshots
keep 1 snapshots:
ID        Time                 Host              Tags        Reasons           Paths
----------------------------------------------------------------------------------------------
Y         2023-07-25 12:58:48  <host>                        daily snapshot    /backup.vma.zst
                                                             weekly snapshot
                                                             monthly snapshot
                                                             yearly snapshot
----------------------------------------------------------------------------------------------
1 snapshots

remove 1 snapshots:
ID        Time                 Host              Tags        Paths
----------------------------------------------------------------------------
X         2023-07-25 12:12:29  <host>                        /backup.vma.zst
----------------------------------------------------------------------------
1 snapshots

[0:00] 100.00%  1 / 1 files deleted
1 snapshots have been removed, running prune
loading indexes...
loading all snapshots...
finding data that is still in use for 1 snapshots
[0:00] 100.00%  1 / 1 snapshots
searching used packs...
collecting packs for deletion and repacking
[0:00] 100.00%  5194 / 5194 packs processed

to repack:             0 blobs / 0 B
this removes:          0 blobs / 0 B
to delete:         29531 blobs / 43.232 GiB
total prune:       29531 blobs / 43.232 GiB
remaining:         29513 blobs / 43.263 GiB
unused size after prune: 15.433 MiB (0.03% of remaining size)

rebuilding index
[0:01] 100.00%  2600 / 2600 packs processed
deleting obsolete index files
[0:00] 100.00%  10 / 10 files deleted
removing 2594 old packs
[0:06] 100.00%  2594 / 2594 files deleted
done
```

After this ran, my expectation was that the bucket size was going to be cut in half as I had removed one of the snapshots, however, after waiting and waiting for Backblaze's UI to update, neither the bucket size or the file count changed at all, despite the forget-prune operation reporting "to delete: 29531 blobs / 43.232 GiB".

I even ran `restic stats --mode raw-data` and it reports the size to be half of what it is reported by Backblaze, so Restic really seems to think the data has been removed from the remote, but according to Backblaze, it hasn't.

Here's the env vars that were present when both backup and forget commands were ran:
```bash
RESTIC_REPOSITORY='s3:s3.eu-central-003.backblazeb2.com/<bucket_name>'
RESTIC_PASSWORD='REDACTED'

AWS_ACCESS_KEY_ID='REDACTED'
AWS_SECRET_ACCESS_KEY='REDACTED'

KEEP_DAILY='1'
KEEP_WEEKLY='1'
KEEP_MONTHLY='1'
KEEP_YEARLY='1'
```

And here's my Backblaze bucket config:
```
Type: Private
File Lifecycle: Keep only the last version
Snapshots: 0
Encryption: Disabled
```

What backend/server/service did you use to store the repository?
----------------------------------------------------------------
Backblaze B2 (via S3).

Expected behavior
-----------------
Backblaze B2 bucket size to be reduced to 1 snapshots' worth of data (around 50GB) and file count to lower.

Actual behavior
---------------
Backblaze B2 bucket size and file count did not change.

Do you have any idea what may have caused this?
-----------------------------------------------
I haven't seen this behaviour with other repositories I use `restic` in. Could it have to do with the type of data I'm attempting to backup, or the way I'm doing it (stdout to stdin) ?

Do you have an idea how to solve the issue?
-------------------------------------------
No :(

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------
Yes! I love Restic. Along with a few scripts I wrote to manage the entire operation, it has changed my backup workflow massively (for the better ofc) <3

**Comment by MichaelEischer** [comment]
as you're using the S3 backend, the actual deletion of the file content happens via a lifecycle rule. According to https://www.backblaze.com/docs/cloud-storage-lifecycle-rules for "Keep only the last version":

> The previous version of the file is “hidden” for one day and then deleted.

The S3 backend when used on B2 delete files by marking them as hidden. After one day the remainder should be cleaned up by the lifecycle rule. That is, if you recheck the bucket size now, it should only contain about 43 GB.

**Comment by hugoamvieira** [comment]
> as you're using the S3 backend, the actual deletion of the file content happens via a lifecycle rule. According to https://www.backblaze.com/docs/cloud-storage-lifecycle-rules for "Keep only the last version":
> 
> > The previous version of the file is “hidden” for one day and then deleted.
> 
> The S3 backend when used on B2 delete files by marking them as hidden. After one day the remainder should be cleaned up by the lifecycle rule. That is, if you recheck the bucket size now, it should only contain about 43 GB.

This makes a lot of sense, thanks for bringing it up 😄 - I will try and do a longer test to see if the Bucket size reduces after the day!

**Comment by hugoamvieira** [comment]
Update:

I deleted all the data from the bucket, re-init it with restic and performed a backup (1 snapshot).
After this, B2 reported:

```
Current Files: 2,395
Current Size: 42.6 GB
```

A bit later (01/08/23), I performed another backup operation followed by a forget-prune operation which deleted the first restic snapshot. The entire process finished that same day at 11:24am.

Right after that op, there was 1 snapshot present as reported by `restic snapshots`, and `Total Size: 43.523 GiB`, as reported by `restic stats --mode raw-data`.

B2 reported (After 2nd backup & forget-prune - Finished at 01/08/23 11:24am GMT+1):
```
Current Files: 7381
Current Size: 89.1 GB
```

So far, so good. I then waited a full day and checked B2 again (02/08/23 12:11am GMT+1):
```
Current Files: 7379
Current Size: 89.1 GB
```

So unfortunately there was no change on bucket size and 2 less files. `restic stats --mode raw-data` also reported the same `Total Size: 43.523 GiB`.

A weird tidbit is that, despite having the bucket set to only keep the last version, I still see a lot of versioned files (with `(2) *` in the file name) in the bucket, even after the day wait -- My expectation here was that they'd get deleted by Backblaze after the one day as per the docs?

I tried getting some info using the B2 CLI by hitting the [get-file-info](https://www.backblaze.com/apidocs/b2-get-file-info) API, but I found no reason for these file versions to be there.

I also checked the bucket properties, and see nothing out of the ordinary:
```json
    "defaultRetention": {
        "mode": "unknown"
    },
    "isFileLockEnabled": null,
    "lifecycleRules": [
        {
            "daysFromHidingToDeleting": 1,
            "daysFromUploadingToHiding": null,
            "fileNamePrefix": ""
        }
    ],
    "options": [
        "s3"
    ],
    "replication": {
        "asReplicationDestination": null,
        "asReplicationSource": null
    },
```

Particularly, the `lifecycleRules` are consistent with the "Keep only the last version" JSON in the documentation. This same API also corroborates the file count and size seen on the dashboard.

I'm going to wait until tomorrow to see if the files eventually make their way out and report back - There may be some timing sutff at play here.

I also thought of using the [B2 backend directly](https://restic.readthedocs.io/en/latest/030_preparing_a_new_repo.html#backblaze-b2) instead of the S3 API, just to see if the behaviour is different, however, as per the warning in the docs, even if it worked better, I don't know how advisable this is to use in an actual backup scenario.
